### DISCUSSION TITLE
Stable server actions

### DISCUSSION THREAD

**Discussion by MZHoffman** [vercel/next.js#52107] [Help]
### Summary

Hello all!

How do you think, when stable actions will be stable.
I'm working on a project and thinking of using them but not sure if it's a good idea now.

Cheers!

**Comment by chrislgrigg** [comment]
I'm also interested in this. If I may piggyback on it, I'd like to ask a related question: is the recommendation to not use Server Actions in prod coming from any specific known problems/limitations or more of a general concern about how new it is?

**Comment by sinafath** [comment]
yeah we are using cookies and we don't know how to handle it without next server actions

**Comment by domwakeling** [comment]
Given the emphasis on server actions, eg the current [deep dive discussion](https://github.com/vercel/next.js/discussions/54075) it seems clear they're the future - and the additional functionality (simlply for data invalidation/revalidation) is significant.

But

As others above, and elsewhere, are pointing out — reluctant to start adopting them without some indication of how long it may be before they **might** be considered stable. I'm facing down either providing a sub-optimal experience for a client, or implementing server actions and taking the (perceived? actual?) risk of additional work down the line when the API or some other implementation details change.

**Reply by domwakeling** [reply]
(or, third option, turn off data cahing and dramatically increasing database utilisation/cost — as you can imagine, not a popular choice)

**Reply by chrislgrigg** [reply]
> I'm facing down either providing a sub-optimal experience for a client, or implementing server actions and taking the (perceived? actual?) risk of additional work

@domwakeling After asking through a couple channels and getting no response, I’ve decided to just YOLO and launch a personal project that uses Server Actions. The basics seem perfectly stable so far when served by a Node.js process, I can’t say for sure how it’ll work in Serverless world. I’m choosing to take the lack of updates to mean that they’re working well enough. Time will tell if this bet will pay off.

**Comment by abdullah-wn** [comment]
Hi,

Server actions aren't a feature created by Vercel in Next.js, but rather a feature (similar to server components) inherent to React itself. You can find some implementations in the React codebase at:

[https://github.com/facebook/react/tree/main/packages/react-server/src](https://github.com/facebook/react/tree/main/packages/react-server/src)

Also, you can see that the codebase is from the "react-server" package, and this statement can be found in the readme:

"This is an **experimental package** for creating custom React streaming server renderers.

**Its API is not as stable** as that of React, React Native, or React DOM, and it does not follow the common versioning scheme.

**Use it at your own risk.**"

This is why the Next.js team mentioned in their blog:

"[Server Actions](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions) are available today in alpha with Next.js 13.4. By opting into using Server Actions, **Next.js will use the experimental release channel of React.**"

[Source](https://nextjs.org/blog/next-13-4#turbopack-beta)

Server actions are built on top of a React experimental feature, which isn't documented yet. I have found a official communication from the React team regarding this feature in this post :

https://react.dev/blog/2023/03/22/react-labs-what-we-have-been-working-on-march-2023#react-server-components

Also i found this in their blog:

"Experimental releases may be significantly different than releases to Canary and Latest. **Do not use Experimental releases in user-facing applications.** You should expect frequent breaking changes between releases in the Experimental channel."

[Source](https://react.dev/community/versioning-policy#experimental-channel)

And this :
"Why not use experimental releases instead? 

Although you can technically use [Experimental releases](https://react.dev/community/versioning-policy#canary-channel), we recommend against using them in production because experimental APIs can undergo significant breaking changes on their way to stabilization (or can even be removed entirely)."

[Source](https://react.dev/blog/2023/05/03/react-canaries#why-not-use-experimental-releases-instead)

For now, it absolutely can't be considered a stable feature, but there might be developments in the near future.

So, it's not stated as stable yet.

Don’t use server actions in a real project

It’s a very early feature
